### PR TITLE
36/fix leftovers problem

### PR DESCRIPTION
**Pull request by RandSohail**
### To be fixed:
- [x] #37
- [x] unify the sent message to the user when signup to be `user does not exist` instead of exposing his/her current status `PENDING | REJECT | APPROVED` in `registrationChecks`
- [x] handle reason of rejection coming from the front to `api/v1/user/reject/:id`
- [x] handle waiting list pagination
- [x] validate the email to follow a regex & handle the validation error to be sent to the front through a msg.
- [x] change the sent message when the user enter their already approved email to be something else than that `CHECK EMAIL` to distinguish it from other messages.
- [ ] handle this error occurred Google login/signup process
    
- [x] create a controller for changing passwords.

### CODE REVIEW

**Review by Ibrahim-Jarada — CHANGES_REQUESTED** [review]
Good Job in general!

File: src/middleware/checkUserRole.ts [hunk 1]
```diff
@@ -1,37 +1,39 @@
...
     const { userRoleId, userStatusId } = userData;
 
-    if (!userTypes.includes(userRoleId) || userStatusId !== USER_STATUS.APPROVED) {
-      throw new CustomError(UNAUTHORIZED, constants.HttpStatus.UNAUTHORIZED);
+    if (!userTypes.includes(userRoleId) || userStatusId !== userStatus.APPROVED) {
+      throw errorMessages.UNAUTHORIZED_ERROR;
     }
 
-    req.user = userData;
+    req.app.set('user', userData);
```
Comment (Ibrahim-Jarada) [inline src/middleware/checkUserRole.ts:36]: !Important, This is a global setting as we use a lot of async functions users requests may affect others ones and that will lead to unexpected behavior (high severity vulnerability)

**Review by Ibrahim-Jarada — CHANGES_REQUESTED** [review]
Security Problem

**Review by zaher-aa — CHANGES_REQUESTED** [review]

**Review by Ibrahim-Jarada — APPROVED** [review]

**Review by zaher-aa — APPROVED** [review]

**Comment by zaher-aa** [comment]
**DO NOT** merge this pull-request until refactoring the front end side, so the endpoints we use in the front end match the ones we have in our API

File: src/helpers/errorMessages.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import CustomError from './CustomError';
+import { messages, httpStatus } from './constants';
+
+export const NOT_EXIST_ERROR = new CustomError(
```
Comment (Jood80) [inline src/helpers/errorMessages.ts:4]: Nice move!!
Optionally, as you're already calling them as `errroMesseges.NOT_EXIST_ERROR` I would suggest to remove the `_ERROR` word

File: src/helpers/validation/createUserSchema.ts [hunk 3]
```diff
@@ -1,7 +1,8 @@
 import Joi from 'joi';
+import { name, email, idValidation } from '../validationRules';
 
 export default Joi.object({
-  name: Joi.string().required(),
-  email: Joi.string().email().required(),
-  roleId: Joi.number().required(),
```
Comment (Jood80) [inline src/helpers/validation/createUserSchema.ts:6]: I would suggest to add an extra layer of validation here, say:
```
Joi.number().valid(1, 2, 3).isRequired()
```
for instance

File: src/helpers/validationRules.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+// export const date = Joi.string().isoDate().required();
+
+// export const guid = Joi.string().guid();
+
+// export const passwordRef = Joi.string().invalid(Joi.ref('oldPassword'))
+// .messages({ 'any.invalid': validationRulesMessages.PASSWORD_MATCH }).required();
+
+// export const confirmPassword = Joi.ref('password');
+
+// export const uri = Joi.string().uri().required();
```
Comment (Jood80) [inline src/helpers/validationRules.ts:80]: what's the purpose of these commented lines?

File: src/middleware/checkUserRole.ts [hunk 5]
```diff
@@ -1,38 +1,40 @@
...
-  constants, CustomError, verifyToken, tokenError,
+  constants, verifyToken, tokenError, errorMessages,
 } from '../helpers';
 import { getAllUserDataById } from '../services';
+import { IUserRequest } from '../interfaces';
 
 export default (userTypes: number[]) => async (
-  req: UserAuth,
+  request: IUserRequest,
   _res: Response,
```
Comment (Jood80) [inline src/middleware/checkUserRole.ts:10]: you're naming them one of them `request` and the other one `_res`, make up your mind 👓

File: src/middleware/checkUserRole.ts [hunk 6]
```diff
@@ -1,38 +1,40 @@
...
   try {
-    const { accessToken } = req.cookies;
+    const { accessToken } = request.cookies;
 
-    if (!accessToken) throw new CustomError(UNAUTHORIZED, constants.HttpStatus.UNAUTHORIZED);
+    if (!accessToken) {
+      throw errorMessages.UNAUTHORIZED_ERROR;
+    }
 
     const userPayload = await verifyToken(accessToken);
```
Comment (Jood80) [inline src/middleware/checkUserRole.ts:21]:
```suggestion
    const { id } = await verifyToken(accessToken);
```

File: src/routes/contents/index.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+} from '../../helpers';
+import { checkUserRole } from '../../middleware';
+
+const router = Router();
+
+const { ADMIN, MASTER_ADMIN } = constants.userRoles;
+
+router.use(checkUserRole([ADMIN, MASTER_ADMIN]));
+
+router.get('/un-match-contents', validator.query(getPaginatedDataSchema), getUnMatchContent);
```
Comment (Jood80) [inline src/routes/contents/index.ts:21]: Preferably: `/unmatched-content`

File: src/services/contents/getUnMatchContents.ts [hunk 8]
```diff
@@ -37,4 +32,4 @@ const getPaginatedContents: IGetPaginatedContents = ({
   });
 };
 
-export default getPaginatedContents;
+export default getUnmatchedContent;
```
Comment (Jood80) [inline src/services/contents/getUnMatchContents.ts:35]: change it's file name too, to `getUnmatchedContent` instead of `getUnMatchContents`

**Review by Jood80 — APPROVED** [review]
such a great job you've done here, Rand!! 🔥 
please give the comments I left to you a look

**Comment by Jood80** [comment]
And please solve the conflicts too

**Pull request merged**